src/System.Net.HttpListener/src/System/Net/Managed/WebSockets/WebSocketValidate.Http.Managed.cs

```diff
@@ -0,0 +1,112 @@
+
+            string secWebSocketVersion = request.Headers[HttpKnownHeaderNames.SecWebSocketVersion];
+
+            // Optional for non-browser client
+            string origin = request.Headers[HttpKnownHeaderNames.Origin];
+
+            List<string> secWebSocketProtocols = new List<string>();
+            string outgoingSecWebSocketProtocolString;
+            bool shouldSendSecWebSocketProtocolHeader =
+                ProcessWebSocketProtocolHeader(
+                    request.Headers[HttpKnownHeaderNames.SecWebSocketProtocol],
+                    subProtocol,
+                    out outgoingSecWebSocketProtocolString);
+
+            if (shouldSendSecWebSocketProtocolHeader)
+            {
+                secWebSocketProtocols.Add(outgoingSecWebSocketProtocolString);
+                response.Headers.Add(HttpKnownHeaderNames.SecWebSocketProtocol,
+                    outgoingSecWebSocketProtocolString);
+            }
+
+            // negotiate the websocket key return value
+            string secWebSocketKey = request.Headers[HttpKnownHeaderNames.SecWebSocketKey];
+            string secWebSocketAccept = WebSocketValidate.GetSecWebSocketAcceptString(secWebSocketKey);
+
+            response.Headers.Add(HttpKnownHeaderNames.Connection, HttpKnownHeaderNames.Upgrade);
+            response.Headers.Add(HttpKnownHeaderNames.Upgrade, WebSocketUpgradeToken);
+            response.Headers.Add(HttpKnownHeaderNames.SecWebSocketAccept, secWebSocketAccept);
+
+            response.StatusCode = (int)HttpStatusCode.SwitchingProtocols; // HTTP 101
+            response.StatusDescription = HttpStatusDescription.Get(HttpStatusCode.SwitchingProtocols);
+
+            HttpResponseStream responseStream = response.OutputStream as HttpResponseStream;
+            
+            // Send websocket handshake headers
+            await responseStream.WriteWebSocketHandshakeHeadersAsync();
+
+            WebSocket webSocket = ManagedWebSocket.CreateFromConnectedStream(context.Connection.ConnectedStream, true, subProtocol, keepAliveInterval, receiveBufferSize, internalBuffer);
+
+            HttpListenerWebSocketContext webSocketContext = new HttpListenerWebSocketContext(
+                                                                request.Url,
+                                                                request.Headers,
+                                                                request.Cookies,
+                                                                context.User,
+                                                                request.IsAuthenticated,
+                                                                request.IsLocal,
+                                                                request.IsSecureConnection,
+                                                                origin,
+                                                                secWebSocketProtocols.AsReadOnly(),
```

Comment:
Nit: it looks like we're only ever adding a single item... rather than using a ```List<T>```, could we just have an array of one element, or ```Array.Empt<string>``` if there weren't any?